File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        
        # Return lists if either are empty.
        if list1 is None:
            return list2
        elif list2 is None:
            return list1
        
        # Set head.
        head = None
        if list1.val <= list2.val:
            head = list1
            list1 = list1.next
        else:
            head = list2
            list2 = list2.next
        
        # "head is a reference to the first node in your merged list, and tail is also a reference to this same node. 
        # They are not pointing to each other, but rather they are both pointing to the same location 
        # in memory (the first node of your merged list)."
        
        # Set tail.
        tail = head
        
        # Traverse linked list.
        while (list1 is not None) and (list2 is not None):
            temp = None # Temp to store next object.
            if list1.val <= list2.val:
                temp = list1
                list1 = list1.next
            else:
                temp = list2
                list2 = list2.next
            
            tail.next = temp
            tail = temp
        
        # Fill in either remainder of list1 or list2.
        if list1 is not None:
            tail.next = list1
            
        if list2 is not None:
            tail.next = list2
        
        return head

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        
        ListNode head = null;
        
        // Return either lists if one of them is NULL.
        if (list1 == null) {
            return list2;
        } else if (list2 == null) {
            return list1;
        }
        
        // Set either list1 or list2 as the head.
        if (list1.val <= list2.val) {
            head = list1;
            list1 = list1.next;
        } else {
            head = list2;
            list2 = list2.next;
        }
        
        // Set tail.
        ListNode tail = head;
        
        // Iterate through lists.
        while (list1 != null && list2 != null) {
            ListNode tmp = null;
            if (list1.val <= list2.val) {
                tail.next = list1;
                list1 = list1.next;
            } else {
                tail.next = list2;
                list2 = list2.next;
            }
            tail = tail.next;   // Move tail pointer forward.
        }
        
        // Link remaining lists.
        if (list1 != null) {
            tail.next = list1;
        } else if (list2 != null) {
            tail.next = list2;
        }
            
        return head;
    }
}